x/subscription/keeper/keeper.go

```diff
@@ -0,0 +1,405 @@
+	}
+	plan := types.Plan{
+		Owner:        msg.Owner,
+		Title:        msg.Title,
+		Description:  msg.Description,
+		Price:        msg.Price,
+		PlanId:       planID,
+		DurationSecs: msg.DurationSecs,
+		CronSpec:     msg.CronSpec,
+	}
+
+	k.SetPlan(ctx, plan)
+	k.SetPlanID(ctx, planID+1)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeCreatePlan,
+			sdk.NewAttribute(types.AttributeKeyPlanID, strconv.FormatInt(int64(plan.PlanId), 10)),
+		),
+	})
+	return plan, nil
+}
+
+func (k Keeper) StopPlan(ctx sdk.Context, planID uint64) {
+	store := ctx.KVStore(k.storeKey)
+
+	// remove all subscriptions
+	k.IterateSubscriptionsOfPlan(ctx, planID, func(subscription types.Subscription) bool {
+		k.RemoveSubscription(ctx, subscription)
+		return false
+	})
+
+	store.Delete(types.PlanKey(planID))
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeStopPlan,
+			sdk.NewAttribute(types.AttributeKeyPlanID, strconv.FormatInt(int64(planID), 10)),
+		),
+	})
+}
+
+// GetSubscriptionID gets the highest subscription ID
+func (k Keeper) GetSubscriptionID(ctx sdk.Context) (subscriptionID uint64, err error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.SubscriptionIDKey)
+	if bz == nil {
+		return 0, sdkerrors.Wrap(types.ErrInvalidGenesis, "initial subscription ID hasn't been set")
+	}
+
+	subscriptionID = types.GetSubscriptionIDFromBytes(bz)
+	return subscriptionID, nil
+}
+
+// SetSubscriptionID sets the new subscription ID to the store
+func (k Keeper) SetSubscriptionID(ctx sdk.Context, subscriptionID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.SubscriptionIDKey, types.GetSubscriptionIDBytes(subscriptionID))
+}
+
+func (k Keeper) CreateSubscription(ctx sdk.Context, msg types.MsgCreateSubscription) (types.Subscription, error) {
+	plan, success := k.GetPlan(ctx, msg.PlanId)
+	if !success {
+		return types.Subscription{}, errors.New("plan not exists")
+	}
+	blockTime := ctx.BlockTime().Unix()
+	subscriptionID, err := k.GetSubscriptionID(ctx)
+	if err != nil {
+		return types.Subscription{}, err
+	}
+	subscription := types.Subscription{
+		SubscriptionId:     subscriptionID,
+		PlanId:             msg.PlanId,
+		Subscriber:         msg.Subscriber,
+		CreateTime:         uint64(blockTime),
+		NextCollectionTime: uint64(plan.CronSpec.Compile().RoundUp(blockTime+1, plan.Tzoffset)),
+		ExpirationTime:     uint64(blockTime) + uint64(plan.DurationSecs),
+		PaymentFailures:    0,
+	}
+	k.SetSubscription(ctx, subscription)
+	k.SetSubscriptionID(ctx, subscriptionID+1)
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeCreateSubscription,
+			sdk.NewAttribute(types.AttributeKeySubscriptionID, strconv.FormatInt(int64(subscription.SubscriptionId), 10)),
+		),
+	})
+
+	var gasPerCollection uint32
+	k.paramSpace.Get(ctx, types.KeyGasPerCollection, &gasPerCollection)
+	ctx.GasMeter().ConsumeGas(sdk.Gas(gasPerCollection), "create subscription")
```

Comment:
in spec, it says:
```
ConsumeGas( count_period(create_time, expiration_time, plan.cron_spec, plan.cron_tz) * GasPerCollection )
```

but here, it'll just charge a fixed param gasPerCollection?

Reply:
yes, forgot that, added now.